### PR TITLE
[#5549]feat(core):support catalog pre event for Gravitino server

### PR DESCRIPTION
**Pull request by sunxiaojian**
### What changes were proposed in this pull request?

Add catalog pre event to event listener

### Why are the changes needed?

Fixed : https://github.com/apache/gravitino/issues/5549

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

Added unit test

### CODE REVIEW

**Comment by sunxiaojian** [comment]
@FANNG1 PTAL

File: core/src/main/java/org/apache/gravitino/listener/api/event/CreateCatalogPreEvent.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  /**
+   * Constructs an instance of {@code CreateCatalogEvent}, capturing essential details about the
+   * successful creation of a catalog.
+   *
+   * @param user The username of the individual who initiated the catalog creation.
+   * @param identifier The unique identifier of the catalog that was created.
+   * @param createdCatalogInfo The final state of the catalog post-creation.
+   */
+  public CreateCatalogPreEvent(
+      String user, NameIdentifier identifier, CatalogInfo createdCatalogInfo) {
```
Comment (FANNG1) [inline core/src/main/java/org/apache/gravitino/listener/api/event/CreateCatalogPreEvent.java:40]: createdCatalogInfo -> createCatalogRequest?

File: core/src/main/java/org/apache/gravitino/listener/api/event/AlterCatalogPreEvent.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   * @param user The username of the individual responsible for initiating the catalog alteration.
+   * @param identifier The unique identifier of the altered catalog, serving as a clear reference
+   *     point for the catalog in question.
+   * @param catalogChanges An array of {@link CatalogChange} objects representing the specific
+   *     changes applied to the catalog during the alteration process.
+   */
+  public AlterCatalogPreEvent(
+      String user, NameIdentifier identifier, CatalogChange[] catalogChanges) {
+    super(user, identifier);
+    this.catalogChanges = catalogChanges.clone();
```
Comment (FANNG1) [inline core/src/main/java/org/apache/gravitino/listener/api/event/AlterCatalogPreEvent.java:44]: remove clone here? no need to clone in pre event

File: core/src/main/java/org/apache/gravitino/listener/api/event/DropCatalogPreEvent.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.gravitino.listener.api.event;
+
+import org.apache.gravitino.NameIdentifier;
+import org.apache.gravitino.annotation.DeveloperApi;
+
+/** Represents an event that is generated before a catalog is successfully dropped. */
```
Comment (FANNG1) [inline core/src/main/java/org/apache/gravitino/listener/api/event/DropCatalogPreEvent.java:25]: remove `successfully`?

File: docs/gravitino-server-config.md [hunk 4]
```diff
@@ -134,6 +134,7 @@ Gravitino triggers a pre-event before the operation, a post-event after the comp
 | Iceberg REST server table operation | `IcebergCreateTablePreEvent`, `IcebergUpdateTablePreEvent`, `IcebergDropTablePreEvent`, `IcebergLoadTablePreEvent`, `IcebergListTablePreEvent`, `IcebergTableExistsPreEvent`, `IcebergRenameTablePreEvent` | 0.7.0-incubating |
 | Gravitino server table operation    | `CreateTablePreEvent`, `AlterTablePreEvent`, `DropTablePreEvent`, `PurgeTablePreEvent`, `LoadTablePreEvent`, `ListTablePreEvent`                                                                           | 0.8.0-incubating |
 | Gravitino server schema operation   | `CreareSchemaPreEvent`, `AlterSchemaPreEvent`, `DropSchemaPreEvent`, `LoadSchemaPreEvent`, `ListSchemaPreEvent`                                                                                            | 0.8.0-incubating |
+| Gravitino server catalog operation  | `CreareCatalogPreEvent`, `AlterCatalogPreEvent`, `DropCatalogPreEvent`, `LoadCatalogPreEvent`, `ListCatalogPreEvent`                                                                                       |                  |
```
Comment (FANNG1) [inline docs/gravitino-server-config.md:137]: CreareCatalogPreEvent -> CreateCatalogPreEvent?

**Comment by FANNG1** [comment]
@sunxiaojian thanks for your PR, LGTM except minior comments

File: core/src/main/java/org/apache/gravitino/listener/api/event/CreateCatalogPreEvent.java (hunk 1 above)
Comment (FANNG1) [reply core/src/main/java/org/apache/gravitino/listener/api/event/CreateCatalogPreEvent.java:40]: please correct java doc too

**Comment by sunxiaojian** [comment]
> @sunxiaojian thanks for your PR, LGTM except minior comments

@FANNG1 Done, thanks for  your review.

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@sunxiaojian , merged to main, thanks for your contribution!

**Pull request merged**